### ISSUE TITLE
NBT data not properly saving/disappearing at random

### ISSUE THREAD

**Issue by Derpyness2000** [SpongePowered/SpongeForge#2097]
Tested sponge versions: 1.12.2-2611-7.1.0-BETA-3016 3003 3002 2990 and 2790
Forge version: 1.12.2 - 14.23.2.2611
Java version: OpenJDK 64-Bit Server VM, version 1.8.0_151
OS: Linux

Plugins (19): Minecraft, Minecraft Coder Pack, SpongeAPI, SpongeForge, CatClearLag, Eagle Factions, LuckPerms, MMCRestrict, MagiBridge, No Downfall, NuVotifier, Nucleus, Prism, SeriousVote, Total Economy, UniversalMarket, Vanish

Mods (103): Minecraft, Minecraft Coder Pack, Forge Mod Loader, Minecraft Forge, SpongeAPI, SpongeForge, AE2 Stuff, Actually Additions, Advanced Generators, Applied Energistics 2, Aroma1997Core, AromaBackup, AromaBackuprecovery, AutoRegLib, BD Lib, Baubles, Better Builder's Wands, Biomes O' Plenty, Block Armor, Botania, CLibrary, ChickenChunks, Chisel, Chunk Pregenerator, CoFH Core, CoFH World, CodeChicken Lib, Compact Solar Arrays, ComputerCraft, CraftTweaker JEI Support, CraftTweaker2, Cucumber Library, Cyberware, DeepResonance, ElecCore, Extended Crafting, External Tweaker, Extra Cells 2, Extra Utilities 2, Extreme Reactors, FoamFix, FoamFixCore, Gravestone Mod, ICBM-Classic, Industrial Foregoing, IndustrialCraft 2, InstantUnify, Iron Chest, IvToolkit, Just Enough Dimensions, Less Forgiving Void, Loot Bags, LootTweaker, Lost Souls, MTLib, Magical Psi, Mantle, MatterOverdrive, Mekanism, MekanismGenerators, MekanismTools, MmmMmmMmmMmm, Moar Tinkers, Mod Tweaker, OMLib, Open Modular Turrets, OpenBlocks, OpenComputers, OpenComputers (Core), OpenModsLib, OpenModsLib Core, OreExcavation, PlusTiC, Psi, Quick Leaf Decay, Random Things, RandomPatches, Redstone Arsenal, Redstone Flux, Refined Storage, Sampling Performance Profiler, SanAndreasPs Mod Library, SanAndreasPs Player Model, Secret Rooms 5, Shadowfacts' Forgelin, TATW, Techguns, Techguns Core, Tesla Core Lib, Tesla Core Lib Registries, The Lost Cities, Thermal Cultivation, Thermal Dynamics, Thermal Expansion, Thermal Foundation, Tinker I/O, Tinkers Tool Leveling, Tinkers' Addons, Tinkers' Construct, UniDict, WanionLib, Zero CORE, mob_grinding_utils

Upon loading and/or unloading chunks many times NBT data has a chance to be wiped with a rare chance of metadata also disappearing.
I'm unable to figure out how exactly i can trigger it.
Chests and signs that had their NBT data wiped will only render when you look at them and come in range to click on them.
And a logger here shows that the chunk that had some chests was saved 3 times without being loaded in the meanwhile while i wasen't anywhere near those chests and upon the 3rd save the tile entities simply weren't saved anymore.
https://www.dropbox.com/s/9dn4cmucyhgo52n/fml-server-latest.log?dl=0
Check the following moments for the 3 saves:
[20:12:11]
[20:12:56]
[20:13:25]

**Comment by Derpyness2000** [comment]
Found the cause: I:dormantChunkCacheSize in forgeChunkLoading.cfg is not compatible with spongeforge And here is where i managed to find that which turned out to work even though spongeforge wasen't present in this issue: https://github.com/MinecraftForge/MinecraftForge/issues/3325

**Comment by phit** [comment]
not a Sponge issue in the first place, that setting is broken with just Forge

**Comment by Derpyness2000** [comment]
Without spongeforge that settings works just fine

**Comment by gabizou** [comment]
I don’t believe we re implemented the asynchronous chunk loading in sponge common, but I do have a hunch that we are breaking it possibly due to previously missed updates to forges chunk loader. Would have to ask @bloodmc or @Zidane to investigate as I have my plate full. Or maybe @kashike if he’s willling.

**Comment by fhntv24** [comment]
It should happen with Forge without sponge too - they just didn't save nbt at all, and then loaded empty cached nbt. It was fixed in 2497, and another bug that could rewrite entity nbt ( and so [un]dupe items ) in 2587.

But this is still an issue with forge 2655, and latest 7.1 spongeforge.
It's pretty easy to reproduce:
1. disable dormantChunkCacheSize
2. Place some blocks with nbt ( it doesn't do anything with meta )
3. restart server and enable cache. Make chunk dirty ( place/break block, etc )
4. fly out of it's render distance, type /save-all, fly back in.

(Meta could change because mods using TE to store meta)

**tldr;**
Here is commit from forge: https://github.com/MinecraftForge/MinecraftForge/commit/f3273ca0c9f81821ecbf9d97cb9e8f45b1e5003c

As you could see, it changes ChunkProviderServer using patches. 
https://github.com/MinecraftForge/MinecraftForge/blob/f3273ca0c9f81821ecbf9d97cb9e8f45b1e5003c/patches/minecraft/net/minecraft/world/gen/ChunkProviderServer.java.patch#L79
Main thing to look at is bottom. This is forge patch to tick() method:
```
+                        net.minecraftforge.common.ForgeChunkManager.putDormantChunk(ChunkPos.func_77272_a(chunk.field_76635_g, chunk.field_76647_h), chunk);
                         this.func_73242_b(chunk);
                         this.func_73243_a(chunk);
                         this.field_73244_f.remove(olong);
```
This is how forge saves chunk to dormant cache.

But sponge overrides ChunkProviderServer's tick method using mixins ( MixinChunkProviderServer in SpongeCommon ), and because of this forge's patch does basically nothing.
So, I think, problem is pretty simple, and fix should be simple too. I just don't know what to do in this case, as I'm not sponge developer, but this should help you.
(Didn't test this solution, but I checked all other patches and you do not overwrite them. So just calling forge's methods should be OK. And minimum version required by SpongeForge is higher than that patch, so it will be available in 100% of cases )

**Comment by bs2609** [comment]
SpongeForge calls that method via its own mixins, it just does so incorrectly:
https://github.com/SpongePowered/SpongeForge/blob/ffa6cae5ed9fadfe0dbc1993ce123dc8f68ae758/src/main/java/org/spongepowered/mod/mixin/core/world/gen/MixinChunkProviderServer.java#L53-L57

The call to `putDormantChunk` is too late from that location - it should be relocated to match https://github.com/MinecraftForge/MinecraftForge/commit/f3273ca0c9f81821ecbf9d97cb9e8f45b1e5003c#diff-eab6876e2a0f7e3ce76120124440362e.

**Comment by Aaron1011** [comment]
I believe the problem was actually an *extra* call to `putDormantChunk`, which was overwriting any data saved in the previous dormant chunk cache entry. This should be fixed in https://github.com/SpongePowered/SpongeForge/commit/a319151a96aa5d913cf398f6356c35d013c4e34a.
